build/resolve_buck.py

```diff
@@ -0,0 +1,135 @@
+    target_buck_version = ""
+    buck_archive_name = ""
+
+    machine = platform.machine().lower()
+    is_x86 = machine == "x86" or machine == "x86_64" or machine == "amd64"
+    is_arm64 = machine == "arm64"
+
+    if sys.platform.startswith("linux") and is_x86:
+        buck_archive_name = "buck2-x86_64-unknown-linux-gnu.zst"
+        target_buck_version = BUCK_VERSION_LINUX_GNU_X86
+    elif sys.platform.startswith("darwin") and is_arm64:
+        buck_archive_name = "buck2-aarch64-apple-darwin.zst"
+        target_buck_version = BUCK_VERSION_DARWIN_ARM64
+    else:
+        return args.buck2 or "buck2"
+
+    if args.buck2:
+        # If we have an explicit buck2 arg, check the version and warn if
+        # there's a mismatch.
+        ver = buck_util.get_buck2_version(args.buck2)
+        if ver != target_buck_version:
+            print(
+                f'The provided buck2 binary reports version "{ver}", but '
+                f'ExecuTorch needs version "{target_buck_version}". Ensure '
+                "that the correct buck2 version is installed or avoid "
+                "explicitly passing the BUCK2 version to automatically "
+                "download the correct version.",
+                file=sys.stderr,
+            )
+        return args.buck2
+    else:
+        # Look for system buck2 and check version. Note that this can return
+        # None.
+        ver = buck_util.get_buck2_version("buck2")
+        if ver == target_buck_version:
+            # Use system buck2.
+            return "buck2"
+        else:
+            # Download buck2 or used previously cached download.
+
+            cache_dir = Path(args.cache_dir)
+            buck2_local_path = (
+                (cache_dir / f"buck2-{target_buck_version}").absolute().as_posix()
+            )
+            buck2_local_zstd_path = (
+                (cache_dir / f"{buck2_local_path}.zstd").absolute().as_posix()
+            )
+
+            if os.path.isfile(buck2_local_path):
+                return buck2_local_path
+
+            buck2_archive_url = f"https://github.com/facebook/buck2/releases/download/2024-02-15/{buck_archive_name}"
+
+            try:
+                print(f"Downloading buck2 from {buck2_archive_url}...", file=sys.stderr)
+                urllib.request.urlretrieve(buck2_archive_url, buck2_local_zstd_path)
+
+                # Extract and chmod.
+                subprocess.run(
+                    ["zstd", "-d", buck2_local_zstd_path, "-o", buck2_local_path],
+                    check=True,
+                )
+
+                file_stat = os.stat(buck2_local_path)
+                os.chmod(buck2_local_path, file_stat.st_mode | stat.S_IEXEC)
+
+                # Clean up the archive, since we don't need it anymore.
+                os.remove(buck2_local_zstd_path)
+
+                return buck2_local_path
+            except Exception as e:
+                print(f"Failed to download buck2: {e}", file=sys.stderr)
```

Comment:
I think it'd be better to hard-fail in this case, to make the behavior more predictable. Otherwise, depending on the network health, a build might use an incompatible version or a compatible version of buck2.

At a higher level, we should give guarantees to the caller about the return value of this function. Will it always return a buck2 path whose version is compatible with the required version?

Reply:
I've updated to let it throw. That will bubble up as a warning from CMake. I could make it a fatal, if preferred.

As for a guarantee, I would need to get the version hashes for all buck2 releases, which I can do, but may be better to leave for a follow-up.